packages/core/src/config/sources.ts

```diff
@@ -0,0 +1,185 @@
+
+import { AbiEvents, getEvents } from "./abi";
+import { ResolvedConfig } from "./config";
+import { buildFactoryCriteria } from "./factories";
+
+/**
+ * Fix issue with Array.isArray not checking readonly arrays
+ * {@link https://github.com/microsoft/TypeScript/issues/17002}
+ */
+declare global {
+  interface ArrayConstructor {
+    isArray(arg: ReadonlyArray<any> | any): arg is ReadonlyArray<any>;
+  }
+}
+
+/**
+ * There are up to 4 topics in an EVM event
+ *
+ * @todo Change this to a more strict type
+ */
+export type Topics = (Hex | Hex[] | null)[];
+
+export type LogFilterCriteria = {
+  address?: Address | Address[];
+  topics?: Topics;
+};
+
+export type FactoryCriteria = {
+  address: Address;
+  eventSelector: Hex;
+  childAddressLocation: "topic1" | "topic2" | "topic3" | `offset${number}`;
+  topics?: Topics;
+};
+
+type BaseSource = {
+  name: string;
+  network: string;
+  chainId: number;
+  abi: Abi;
+  events: AbiEvents;
+  startBlock: number;
+  endBlock?: number;
+  maxBlockRange?: number;
+};
+
+export type LogFilter = BaseSource & {
+  type: "logFilter";
+  criteria: LogFilterCriteria;
+};
+
+export type Factory = BaseSource & {
+  type: "factory";
+  criteria: FactoryCriteria;
+};
+
+export type Source = LogFilter | Factory;
+
+export const sourceIsLogFilter = (source: Source): source is LogFilter =>
+  source.type === "logFilter";
+
+export const sourceIsFactory = (source: Source): source is Factory =>
+  source.type === "factory";
+
+export const buildSources = ({
+  config,
+}: {
+  config: ResolvedConfig;
+}): Source[] => {
+  const contracts = config.contracts ?? [];
+
+  return contracts
+    .map((contract) => {
+      // Note: should we filter down which indexing functions are available based on the filters
+      const events = getEvents({ abi: contract.abi });
+
+      // Resolve the contract per network, filling in default values where applicable
+      return contract.network
+        .map((networkContract) => {
+          // Note: this is missing config validation for checking if the network is valid
+          const network = config.networks.find(
+            (n) => n.name === networkContract.name
+          )!;
+
+          const resolvedFilter = networkContract.filter ?? contract.filter;
+
+          const topics = resolvedFilter
+            ? buildTopics(contract.abi, resolvedFilter)
+            : undefined;
+
+          const sharedSource = {
+            // constants
+            name: contract.name,
+            abi: contract.abi,
+            network: network.name,
+            chainId: network.chainId,
+            events,
+            // optionally overridden properties
+            startBlock: networkContract.startBlock ?? contract.startBlock ?? 0,
+            endBlock: networkContract.endBlock ?? contract.endBlock,
+            maxBlockRange:
+              networkContract.maxBlockRange ?? contract.maxBlockRange,
+          } as const;
+
+          if ("factory" in contract) {
```

Comment:
I think this won't "fall through" in the case where `factory` is not defined on the base contract but is defined per-network?

Reply:
Good catch